terraform/modules/app-container-definition/main.tf

```
@@ -0,0 +1,49 @@
+variable "name" {
+  type = string
+}
+variable "image" {
+  type = string
+}
+variable "environment_variables" {
+  type = map
+}
+variable "log_group" {
+  type = string
+}
+variable "secrets_from_arns" {
+  type = map
+}
+variable "aws_region" {
+  type = string
+}
+output "value" {
+  value = {
+    "name" : var.name,
+    "image" : var.image,
+    "essential" : true,
+    "environment" : [for key, value in var.environment_variables : { name : key, value : value }],
+    "dependsOn" : [{
+      "containerName" : "envoy",
+      "condition" : "START"
+    }],
```

I think we'll need to parameterise this, otherwise this module is hard-coupled to the envoy one (and we'll shortly have a use case for a task that doesn't use Envoy, when we set up Grafana). Perhaps a map `depends_on` of containerName to condition, something like this?
```suggestion
    "dependsOn" : [for key, value in var.depends_on : { containerName : key, condition : value }],
```

---

I'm inclined to leave this hardcoded for now, on the basis that (at the moment) the intended use case for this module is for the gov.uk apps.

There are things like Grafana which will also need task definitions (as you say), but perhaps they're different / unique / simple enough that we could inline their container definitions when we create their resources, instead of trying to make this module totally generic.

---

I do agree that the coupling between this and the envoy module is unpleasant though 🤔

---

How do you feel about this?

```
    "dependsOn" : (var.depends_on_envoy
      ? [{ "containerName" : "envoy", "condition" : "START" }]
      : []
    ),
```

Where `depends_on_envoy` is a bool with no default, so apps would have to say `depends_on_envoy = true` when calling the module.

The alternatives I can think of have drawbacks:

* `depends_on` variable with a default of `{envoy: "START"}` - having the default makes this module _mainly_ for apps in the mesh, and other things (not in the mesh) would have to say `depends_on: {}` to opt out of envoy, which feels weird.
* `depends_on` variable with no default - means every app in the mesh has to pass the exact same `depends_on: {envoy: "START"}` param, which is an annoying bit of duplication, and feels a bit like an abstraction leak (although to be fair, they have to call the envoy settings module, so perhaps that leak has already leaked).

One other thing we could consider is leaning in to the coupling and combining the app-container-definition module and the envoy-settings module. I think we discussed that and didn't like it, because it's less flexible. I'm not 100% sure the flexibility is a good thing though.

---

That's a good point about the default being a bit awkward either way. I think the way to do it would be the second option (empty default and explicit dependency on envoy). Combining the two modules into an opinionated module for the GOV.UK use case is starting to sound more attractive. I'd be fine with either of those tbh.

---

Implemented as an empty default (so apps will need to be explicit about their dependency on envoy) in e4ee6de.